paastech-cloud/api
# User Authentication and Authorization

## What does this PR do ?
Implements endpoints to register and authenticate user in `UsersModule` and `AuthModule` respectively. Adds authorization which depends on the user table's `isAdmin` attribute.

### How should this be manually tested?

  - Step 1 : launch the app using docker-compose or with `npm start` (see README).
  - Step 2 : send a POST request to `localhost:3000/users` with the following JSON body:
  ```
  {
      "username": "test",
      "email": "test@test.com",
      "password": "Test@12345"
  }
  ```
  A non-admin user will be created (you can change the property `isAdmin` to true in the `create` method of `UsersService` if you want to create an admin user).
  - Step 3 : send the same POST request, but to `localhost:3000/auth` to get an access token (valid for half an hour).
  - Step 4 : use the token as Bearer token in Authorization header to send requests to protected routes (all routes which are not marked with `Public()` decorator.

### TODO
Adding email verification and password reset features with the use of `MailModule`.

### Other changes
The app has custom decorators `@Public()` and `@AdminOnly()` that you can add to controllers to make routes accessible to unauthenticated users or accessible to admin only.
